File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # initialize the output
        output = [0] * len(nums);

        # initialize prefix to 1
        # prefix loop: iterate indices 1 through nums.length-1. For each index:
        ## put prefix in output at current index
        ## update prefix: multiply with nums[index]
        prefix = 1
        for i in range(len(nums)):
            output[i] = prefix
            prefix *= nums[i]

        # initialize postfix to 1
        # postfix loop: iterate indices nums.length-1 through 0. For each index:
        ## put postfix in output at current index (multiply)
        ## update postfix: multiply with nums[index]
        postfix = 1
        for i in range(len(nums)-1, -1, -1):
            output[i] *= postfix
            postfix *= nums[i]
        
        # return output
        return output

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.java
class Solution {
    public int characterReplacement(String s, int k) {
        // initialize longest substring length (acoouting for replacements) to zero
        // initilaize left pointer to zero
        // initialize empty hashmap – representing window
        int longest = 0;
        int left = 0;
        Map<Character, Integer> window = new HashMap<>();

        // iterate through right 0 to s.length-1. for each right pointer index:
        //// push [right] to the hashmap
        //// get most frequency from the window. iterate over values()
        //// while: len(window) - most frequency > k:
        ////// decrement window[left] freq.
        ////// update most frequency
        //// update logest substring length
        for (int right = 0; right < s.length(); right++) {
            window.put(s.charAt(right), window.getOrDefault(s.charAt(right), 0)+1);
            int mostFreq = getMostFrequency(window);

            while ((right-left+1) - mostFreq > k) {
                window.put(s.charAt(left), window.get(s.charAt(left))-1);
                left++;
                mostFreq = getMostFrequency(window);
            }

            longest = Math.max(longest, (right-left+1));
        }

        // return longest substring length
        return longest;
    }

    private int getMostFrequency(Map<Character, Integer> map) {
        int mostFreq = 0;

        for (int value : map.values())
            mostFreq = Math.max(mostFreq, value);
        
        return mostFreq;
    }
}
